content/post/deduplicating_pga_with_apollo.md

```diff
@@ -0,0 +1,387 @@
+(community detection step).
+
+HEAD commits in `PGA` contain 54.5 million files spread across 181,481 projects. 
+In order to extract semantic features from the files, we had 
+to limit ourselves to the programming languages with a functional [Babelfish driver](https://docs.sourced.tech/babelfish/languages). This meant ≈26% of files,
+those written in Python, Java, JavaScript, Ruby or Go.
+Nevertheless, there were still 14.1 million files in our corpus. We extracted
+the following features:
+
+- **identifiers**, such as variable or function names.
+- **literals**, e.g. integer or string constant values.
+- **graphlets**, a feature being composed of the types of UAST node and it's children.
+- **children**, a feature being composed of the node's type and it's [quantized](https://en.wikipedia.org/wiki/Quantization_(signal_processing)) 
+number of children nodes.
+- **uast2seq** - a sequence of UAST node types extracted with depth-first tree traversal of limited length.
+- **node2vec** - a sequence of UAST node types produced from random walks in the tree.
+
+The latter two features are not explicitly defined since they depend on several
+hyperparameters: maximum length of a feature, maximum number of random walks, etc.
+We used the maximum length to 3 and 4 for tree traversals and
+to 2 and 3 for random walks (the output is combined).
+The maximum number of random walks was 2 and the number of random steps was 10.
+Those values hold a trade-off between the vocabulary size and the representativeness.
+
+## Houston, we have a problem
+
+Sadly, it turned out that extracting those features crashed our cluster many, *many* 
+times. We faced several issues:
+
+- Running Spark on [Siva files](https://github.com/src-d/go-siva)
+through [jgit-spark-connector](https://github.com/src-d/jgit-spark-connector)
+led to massive amount of temporary data during feature extraction on the worker 
+pods of Kubernetes, which were further killed without notice by the cluster manager,
+- While converting the data to [Parquet](http://parquet.apache.org/) format to avoid this, 
+the workload imbalance due to [funny repos](https://github.com/Katee/git-bomb) made
+individual tasks last insanely long.
+- Once the conversion was done, applying TF-IDF to 1TB of cached data on disk
+led Spark to spend more time on garbage collection then on the actual work.
+- Finally, after we found a way to apply TF-IDF with manual batching (sic!),
+Babelfish failed to parse a considerable amount of the files.
+
+And that does not even take into account the continuous refactoring and bug detection
+in `src-d/ml`, `src-d/jgit-spark-connector`, `src-d/apollo` that any young project
+faces. We had to adapt rapidly to changes upstream.
+
+Nevertheless, after months of big data war, we managed to overcome the challenges
+and extract 6,194,874 distinct features from 7,869,917 million files, out of 102,114
+projects. The sparsity of that huge matrix was 0.00017, the average number
```

Comment:
0.00017 doesn't say if it's very empty or very full, even if the reader can derive it by context. It should be good to add a note.

Reply:
not sure what to add here, since we already give the equivalent matrix size

Reply:
My point was that `0.00017` sparsity is less informative than `Only 1 element out of 5623 of this matrix was non-zero` because it's not clear if a value near zero is very empty or very full. If we keep this formulation, we should at least call it `density` not `sparsity`, `spasity` is usually defined as the number of null elements over the total number of elements so in this case it is usually given as `0.99983`.

Reply:
aaah okay my bad. changing to density then